README.md
# **Decimal Incrementor** *(VSCode Extension)*

This is based on a fork from https://github.com/boriskrg/vscode-incrementor and it was adapted for GLSL shader development where decimal and unclamped values were needed.

Fixed scales (one, tenth, ten) were replaced by configurable generic scales (small, medium, big). See ChangeLog for more information.

> Increment or decrement just about anything!

![Demo](images/demo-main.gif)

## Contents
* [Features](#features)
  * [Numbers](#numbers)
  * [Enumerators](#enumerators)
  * [Multiple Selections](#multiple-selections)
* [Usage](#usage)
  * [Available Commands](#available-commands)
* [Extension Settings](#extension-settings)
  * [Keybindings](#keybindings)
* [Known Issues](#known-issues)
* [Change Log](#change-log)


## Features

### Numbers

![Numbers](images/demo-number.gif)

Numbers can be incremented or decremented by any value, configurable via `settings.json`. This works with integers, decimals, and negatives alike. The only *real* condition is that it is a **finite** number.

### Enumerators

![Enumerators](images/demo-enumerator.gif)

Enumerators can basically be any kind of text, like a variable or function or command.

**They can only contain letters, numbers and dashes and must start with a letter and can't end with a dash.**

In the extension settings you can add an array of strings that you want to cycle through. *i.e.* `["false", "true"]`

Each array will cycle through the containing strings from beginning to end and also loop back around if you have the option set. Each array is considered a separate enumerator so `"false"` can only become `"true"` and vice versa depending what the array contains.

### Multiple Selections

![Multiple Selections](images/demo-multiple-selections.gif)

Incrementor supports multiple selections, even in the same line. They do not all have to be the same type, so one selection/cursor could be a number and another could be an enumerator.

## Usage

> **TIP:** Incrementing/decrementing can work with one or multiple cursors.<br>If there are no selections Incrementor will use the word under the caret(s) then select them.

For this example we will be incrementing a number.

1) Either select the number you wish to increment or just place the caret inside or next to the number.

2) Then, either press the hotkey that corresponds to the value you wish to increment by or open the Command Palette and use the command `Incrementor: Increment by X`.

3) Congratulations! You just **Incrementored** your first number!

### Available Commands

*In the Command Palette (Cmd+Shift+P)*

* `Incrementor: Increment (Small)`
* `Incrementor: Decrement (Small)`
* `Incrementor: Increment (Medium)`
* `Incrementor: Decrement (Medium)`
* `Incrementor: Increment (Big)`
* `Incrementor: Decrement (Big)`

## Extension Settings

### `incrementor.enabled`

Enables or disables Incrementor.

* **Default:** true
* **Must be:** Boolean

### `incrementor.inc*` and `incrementor.dec*`

Increment a value under the cursor.
There are 3 configurable scales mapped to keyboard shortcuts (see keybindings)

* **Small** defaults to 0.01
* **Medium** defaults to 0.10
* **Big** defaults to 1.00

### `incrementor.decimalPlaces`

The number of decimal places to round incremented/decremented decimal numbers to.<br>*(a value of 0 will disable rounding)*

* **Default:** 0
* **Must be:** 0 to 10, Integer

### `incrementor.loopEnums`

After reaching the end of an Enum set, start back at the beginning.

* **Default:** true
* **Must be:** Boolean

### `incrementor.enums`

An array of arrays, each containing a list of enums to cycle through.

* **Default:** `[["false", "true"], ["let", "const"]]`
* **Must be:**
  ```
  Array >
    Arrays >
      Strings
  ```

### Keybindings

For now, default keybindings aren't being included but these are the ones I use.

```json
{
  "command": "incrementor.incSmall",
  "key": "ctrl+up"
},
{
  "command": "incrementor.decSmall",
  "key": "ctrl+down"
},
{
  "command": "incrementor.incMedium",
  "key": "ctrl+alt+up"
},
{
  "command": "incrementor.decMedium",
  "key": "ctrl+alt+down"
},
{
  "command": "incrementor.incBig",
  "key": "ctrl+alt+cmd+up"
},
{
  "command": "incrementor.decBig",
  "key": "ctrl+alt+cmd+down"
}
```

## Known Issues

* When using the redo command *(e.g. `cmd+shift+z`)* the selections can get a little wonky. I have no idea why it does that but I am looking for a solution.

    Undo works as expected.

## Change Log

See the changelog [here](CHANGELOG.md).


CHANGELOG.md
# Change Log
All notable changes to the "incrementor" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [0.2.0]
- Forked from https://github.com/boriskrg/vscode-incrementor
- Fixed scales (one, tenth, ten) were replaced by configurable generic scales (small, medium, big)
- Limits for each scale were virtually removed


test/extension.test.ts
/* tslint:disable:ter-prefer-arrow-callback */

// The module 'assert' provides assertion methods from node
import * as chai from "chai";
chai.should();

// You can import and use all API from the 'vscode' module
// as well as import your extension to test it
import * as vscode from "vscode";
import * as _ from "lodash";

let vEditor: vscode.TextEditor = undefined;
let vDoc: vscode.TextDocument = undefined;
let vSel: vscode.Selection[] = undefined;
let sel: vscode.Selection;
let text: string;

before(function (done) {
	this.timeout(1000); // A very long environment setup.
	setTimeout(done, 500);
});

describe("Numbers", function () {
	this.slow(110);
	this.timeout(200);

	beforeEach(function () {
		vEditor = vscode.window.activeTextEditor;
		vDoc = vEditor.document;
		vSel = vEditor.selections;
	});

	it("should have 1.00 as selected text", function (done) {
		_.delay(
			(done) => {
				sel = new vscode.Selection(1, 19, 1, 23);
				vEditor.selection = sel;
				text = vDoc.getText(vEditor.selection);

				try {
					text.should.equal("1.00");
					done();
				} catch (error) {
					done(error);
				}
			},
			100,
			done
		);
    });
    
	describe("Small", function () {
		it("should increment 1.00 to 1.01", function (done) {
			vscode.commands.executeCommand("incrementor.incSmall");
			_.delay(
				(done) => {
					text = vDoc.getText(vEditor.selection);

					try {
						text.should.equal("1.01");
						done();
					} catch (error) {
						done(error);
					}
				},
				100,
				done
			);
		});

		it("should decrement 1.01 to 1.00", function (done) {
			vscode.commands.executeCommand("incrementor.decSmall");
			_.delay(
				(done) => {
					text = vDoc.getText(vEditor.selection);

					try {
						text.should.equal("1.00");
						done();
					} catch (error) {
						done(error);
					}
				},
				100,
				done
			);
		});

    });
    
    describe("Medium", function () {
		it("should increment 1.00 to 1.10", function (done) {
			vscode.commands.executeCommand("incrementor.incMedium");
			_.delay(
				(done) => {
					text = vDoc.getText(vEditor.selection);

					try {
						text.should.equal("1.10");
						done();
					} catch (error) {
						done(error);
					}
				},
				100,
				done
			);
		});

		it("should decrement 1.10 to 1.00", function (done) {
			vscode.commands.executeCommand("incrementor.decMedium");
			_.delay(
				(done) => {
					text = vDoc.getText(vEditor.selection);

					try {
						text.should.equal("1.00");
						done();
					} catch (error) {
						done(error);
					}
				},
				100,
				done
			);
		});
    });
    
    describe("Big", function () {
		it("should increment 1.00 to 2.00", function (done) {
			vscode.commands.executeCommand("incrementor.incBig");
			_.delay(
				(done) => {
					text = vDoc.getText(vEditor.selection);

					try {
						text.should.equal("2.00");
						done();
					} catch (error) {
						done(error);
					}
				},
				100,
				done
			);
		});

		it("should decrement 2.00 to 1.00", function (done) {
			vscode.commands.executeCommand("incrementor.decBig");
			_.delay(
				(done) => {
					text = vDoc.getText(vEditor.selection);

					try {
						text.should.equal("1.00");
						done();
					} catch (error) {
						done(error);
					}
				},
				100,
				done
			);
		});
	});
});


src/extension.ts
'use strict';
import * as vscode from 'vscode';
import Range = vscode.Range;
import Selection = vscode.Selection;
import Position = vscode.Position;
import TextEditor = vscode.TextEditor;
import TextEditorEdit = vscode.TextEditorEdit;
import TextDocument = vscode.TextDocument;
// import TextLine = vscode.TextLine;

import * as _ from 'lodash';
import BigNumber from 'bignumber.js';

export function activate(context: vscode.ExtensionContext) {

    const inc = new Incrementor();

    const comIncSmall = vscode.commands.registerCommand('incrementor.incSmall', () => {
        inc.run(inc.action.incSmall);
    });

    const comDecSmall = vscode.commands.registerCommand('incrementor.decSmall', () => {
        inc.run(inc.action.decSmall);
    });

    const comIncMedium = vscode.commands.registerCommand('incrementor.incMedium', () => {
        inc.run(inc.action.incMedium);
    });

    const comDecMedium = vscode.commands.registerCommand('incrementor.decMedium', () => {
        inc.run(inc.action.decMedium);
    });


    const comIncBig = vscode.commands.registerCommand('incrementor.incBig', () => {
        inc.run(inc.action.incBig);
    });

    const comDecBig = vscode.commands.registerCommand('incrementor.decBig', () => {
        inc.run(inc.action.decBig);
    });

    context.subscriptions.push(comIncMedium);
    context.subscriptions.push(comDecMedium);
    context.subscriptions.push(comIncSmall);
    context.subscriptions.push(comDecSmall);
    context.subscriptions.push(comIncBig);
    context.subscriptions.push(comDecBig);
}

/**
 * Extension Class
 */
export class Incrementor {
    /**
     * List of RegEx patterns that match different types of strings.
     *
     * @readonly
     */
    private regex = {
        number:  /^(-?\d+\.?\d*)([a-zA-Z%]*)$/,
        enum:    /^(?:[\w]+-)*[\w]+$/,
        rgb:     /^(rgba(?=\((?:[\s\d]+,){3})|rgb(?!\((?:[\s\d]+,){3}))\(\s*(2[0-5]{2}|1\d{2}|[1-9]\d|\d),\s*(2[0-5]{2}|1\d{2}|[1-9]\d|\d),\s*(2[0-5]{2}|1\d{2}|[1-9]\d|\d)\s*(?:\)|(?:,\s*(1|0?\.\d+|0)\)))$/,
        rgbLine: /(rgba(?=\((?:[\s\d]+,){3})|rgb(?!\((?:[\s\d]+,){3}))\(\s*(2[0-5]{2}|1\d{2}|[1-9]\d|\d),\s*(2[0-5]{2}|1\d{2}|[1-9]\d|\d),\s*(2[0-5]{2}|1\d{2}|[1-9]\d|\d)\s*(?:\)|(?:,\s*(1|0?\.\d+|0)\)))/,
        hex:     /^#(?:([A-Fa-f0-9]{2})([A-Fa-f0-9]{2})([A-Fa-f0-9]{2})|([A-Fa-f0-9]{3}))$/,
    };

    private vEditor:       TextEditor;
    private edit:          TextEditorEdit;
    private vDoc:          TextDocument;
    private vSel:          Selection[];
    private delta:         number;
    private wordRange:     Range;
    private isReversed:    boolean;
    private wordString:    string;
    private lastRange:     Range;
    private lastValueDiff: number;
    private replaceRanges: Selection[];
    private settings     = vscode.workspace.getConfiguration('incrementor');
    public action = {
        incSmall: 0.01,
        decSmall: -0.01,
        incMedium:   0.10,
        decMedium:   -0.10,
        incBig:   1.00,
        decBig:   -1.00
    };
    private hiddenSels: {
        isOn: boolean;
        pos: Position;
        diff: number;
    };

    // constructor() {
    //     this.updateSettings();
    // }

    // TODO: Consider moving to just functions (not a class) for easier testing in mocha

    /**
     * Run when a command is triggered from the editor.
     * All inc/dec methods are triggered from this method.
     *
     * @param delta The amount to change by.
     *
     *  _(taken from the `this.action` object)_
     */
    run(delta: number) {
        BigNumber.config({ ERRORS: false });

        const action = _.findKey(this.action, (o) => {
            return o === delta;
        });

        this.updateSettings();

        delta = this.action[action];

        this.delta = delta;
        this.lastRange = undefined;
        this.lastValueDiff = undefined;
        this.hiddenSels = {
            isOn: false,
            pos: undefined,
            diff: 0
        };
        this.replaceRanges = [];

        // Iterate through selections.
        this.vEditor.edit((edit) => {
            this.edit = edit;

            for (const sel of this.vSel) {
                this.wordRange = (sel.isEmpty) ? this.vDoc.getWordRangeAtPosition(sel.active) : sel;
                this.wordString = this.vDoc.getText(this.wordRange);
                this.isReversed = (sel.isEmpty) ? false : sel.isReversed;

                if (this.wordRange && !this.wordRange.isEmpty) {
                    this.changeNumber() ||
                        this.changeEnum() ||
                        this.doNothing(sel);
                }
            }
        });

        this.vEditor.selections = this.replaceRanges;
    }

    private updateSettings() {
        this.vEditor = vscode.window.activeTextEditor;
        this.vDoc = this.vEditor.document;
        this.vSel = this.vEditor.selections;
        this.settings = vscode.workspace.getConfiguration('incrementor');

        this.action = {
			incSmall: (/^(?:[01](?:\.0)?|0\.[1-9])$/.test(this.settings['incSmallValue'])) ? this.settings['incSmallValue'] : 0.01,
			decSmall: (/^(?:[01](?:\.0)?|0\.[1-9])$/.test(this.settings['decSmallValue'])) ? this.settings['decSmallValue'] : -0.01,
			incMedium: this.settings["incMediumValue"] || 0.10,
			decMedium: this.settings["decMediumValue"] || -0.10,
			incBig: this.settings["incBigValue"] || 1.00,
			decBig: this.settings["decBigValue"] || -1.00,
		};
    }

    /**
     * Increments or decrements a number.
     *
     * @private
     * @param editor TextEditorEdit object passed through to make edits.
     * @returns Success_(true)_ or fail_(false)_.
     */
    private changeNumber(): boolean {
        if (this.regex.number.test(this.wordString)) {
            const prevChar = this.getPrevChar();
            let tempString = this.wordString;
            let tempRange  = this.wordRange;

            if (prevChar.char === '-' && tempString[0] !== '-') {
                tempRange = new Range(prevChar.pos, this.wordRange.end);
                tempString = this.vDoc.getText(tempRange);
            }

            // [0] = Original string
            // [1] = Number part
            // [2] = String part (if any)
            const wordReg = this.regex.number.exec(tempString);
            let partNumber = new BigNumber(wordReg[1]);
            const partText = (wordReg[2]) ? wordReg[2] : '';

            if (partNumber.isFinite()) {

                partNumber = partNumber.plus(this.delta);

                const decPlaces = new BigNumber(this.settings['decimalPlaces']).absoluteValue();

                // decPlaces = 0 === rounding off
                if (decPlaces.isInteger() && decPlaces.greaterThan(0)) {
                    partNumber = partNumber.round(decPlaces.toNumber());
                }
                
                let fixed
                if (wordReg[1].indexOf('.') !== -1) {
                    fixed = wordReg[1].length - wordReg[1].indexOf('.') - 1
                } else if (partNumber.toString().indexOf('.') !== -1) {
                    let _strPN = partNumber.toString()
                    fixed = _strPN.length - _strPN.indexOf('.') - 1
                } else {
                    fixed = 0
                }
                const wordChanged = partNumber.toFixed(fixed) + partText;
//                 const wordChanged = partNumber.toString() + partText;

                return this.replace(tempRange, wordChanged);
            } else {
                return false;
            }
        } else {
            return false;
        }
    }

    private changeEnum() {
        if (this.regex.enum.test(this.wordString) && this.settings.has('enums')) {
            let prevChar   = (this.wordRange.start.character > 0) ? this.getPrevChar() : undefined;
            let tempRange  = this.wordRange;
            let tempString = this.wordString;

            let loopStart = true;
            while (loopStart) {
                if (prevChar && prevChar.char === '-') {
                    prevChar = this.getPrevChar(prevChar.pos.translate(0, -1));

                    while (/\w/.test(prevChar.char) && prevChar.pos.character > 0) {
                        prevChar = this.getPrevChar(prevChar.pos.translate(0, -1));
                    }

                    tempRange = new Range(prevChar.pos.translate(0, 1), tempRange.end);
                    tempString = this.vDoc.getText(tempRange);
                } else {
                    loopStart = false;
                }
            }

            prevChar = this.getPrevChar(this.wordRange.end);

            let loopEnd = true;
            while (loopEnd) {
                if (prevChar.char === '-') {
                    prevChar = this.getPrevChar(prevChar.pos.translate(0, 1));

                    while (/\w/.test(prevChar.char)) {
                        prevChar = this.getPrevChar(prevChar.pos.translate(0, 1));
                    }

                    tempRange = new Range(tempRange.start, prevChar.pos);
                    tempString = this.vDoc.getText(tempRange);
                } else {
                    loopEnd = false;
                }
            }

            let wordChanged = tempString;
            for (const enums of this.settings['enums']) {
                if (_.includes(enums, tempString)) {
                    const eIndex = _.indexOf(enums, tempString);

                    // Cycle enums
                    if (this.delta === this.action.incMedium) {
                        if (enums.length - 1 === eIndex && this.settings['loopEnums']) {
                            // Is last, Cycle around
                            wordChanged = enums[0];
                        } else if (enums.length - 1 === eIndex) {
                            // Is last, don't inc
                            wordChanged = enums[eIndex];
                        } else {
                            wordChanged = enums[eIndex + 1];
                        }
                    } else if (this.delta === this.action.decMedium) {
                        if (eIndex === 0 && this.settings['loopEnums']) {
                            // Is last, Cycle around
                            wordChanged = enums[enums.length - 1];
                        } else if (eIndex === 0) {
                            // Is last, don't inc
                            wordChanged = enums[0];
                        } else {
                            wordChanged = enums[eIndex - 1];
                        }
                    }

                    return this.replace(tempRange, wordChanged);
                }
            }

            return false;
        } else {
            return false;
        }
    }

    /*private changeRGB(editor: TextEditorEdit) {
        let exps: {range: Range, text: string} = this.checkLineForPattern(this.wordRange.start, this.regex.rgbLine);

        if (exps) {
            if (_.isArray(exps)) {
                for (const exp of exps as any[]) {
                    if (exp.range.contains(this.wordRange)) {
                        exps = exp;
                        break;
                    }
                }
            } else {
                if (!exps.range.contains(this.wordRange)) {
                    return false;
                }
            }

            // const expsSplitRGB = exps.text.split(/(\d+)(?!\s*\)|\.)/);
            const expsSplitAlpha = (/(1|0?\.\d+|0)(?=\s*\))/.test(exps.text)) ? exps.text.split(/(1|0?\.\d+|0)(?=\s*\))/) : false;
            // const rangeR = new Range(exps.range.start.translate(0, expsSplitRGB[0].length), exps.range.start.translate(0, expsSplitRGB[0].length + expsSplitRGB[1].length));
            // const rangeG = new Range(rangeR.end.translate(0, expsSplitRGB[2].length), rangeR.end.translate(0, expsSplitRGB[2].length + expsSplitRGB[3].length));
            // const rangeB = new Range(rangeG.end.translate(0, expsSplitRGB[4].length), rangeG.end.translate(0, expsSplitRGB[4].length + expsSplitRGB[5].length));

            const rangeA = (expsSplitAlpha !== false) ? new Range(exps.range.start.translate(0, expsSplitAlpha[0].length), exps.range.start.translate(0, expsSplitAlpha[0].length + expsSplitAlpha[1].length)) : false;

            let tempNumber = new BigNumber(this.wordString);
            if (rangeA && this.wordRange.isEqual(rangeA)) {
                if (this.delta === this.action.incSmall || this.delta === this.action.decSmall) {
                    tempNumber = new BigNumber(_.clamp(tempNumber.plus(this.delta).toNumber(), 0, 1));
                }
            } else {
                if (this.delta !== this.action.incSmall || this.delta !== this.action.decSmall) {
                    tempNumber = new BigNumber(_.clamp(tempNumber.plus(this.delta).toNumber(), 0, 255));
                }
            }

            editor.replace(this.wordRange, tempNumber.toString());
            this.replaceRanges.push(new Selection(this.wordRange.start, this.wordRange.start.translate(0, tempNumber.toString().length)));
        } else {
            return false;
        }
    }*/

    private doNothing(sel: Selection) {
        this.replaceRanges.push(sel);

        return true;
    }

    private getPrevChar(pos?: Position) {
        if (this.wordRange.start.character === 0) {
            return {
                pos,
                char: ''
            };
        }

        if (pos === undefined && this.wordRange.start.character > 0) {
            pos = this.wordRange.start.translate(0, -1);
        }

        return {
            pos,
            char: this.vDoc.getText(new Range(pos, pos.translate(0, 1)))
        };
    }

    /**
     * Checks if there are any Selections after `pos`.
     *
     * @private
     * @param pos The Position to look after.
     * @returns `true` if Selections exist after `pos`, otherwise `false`.
     */
    private hasSelectionsAfter(pos: Position) {
        for (const sel of this.replaceRanges) {
            if (sel.start.isAfter(pos)) {
                return true;
            }
        }

        return false;
    }

    private hasHiddenSelsAfter(pos: Position) {
        let i = 0;
        for (const sel of this.vSel) {
            if (!this.replaceRanges[i] && sel.start.isAfter(pos)) {
                return true;
            }

            i += 1;
        }

        return false;
    }

    /**
     * Shifts all Selections on line after `pos` by `amount`.
     *
     * @private
     * @param pos The Position to look after.
     * @param amount The amount to shift the selection by.
     */
    private shiftSelectionsAfter(pos: Position, amount: number): void {
        this.replaceRanges.forEach((sel, index) => {
            if (sel.start.isAfter(pos)) {
                this.replaceRanges[index] = new Selection(this.replaceRanges[index].start.translate(0, amount), this.replaceRanges[index].end.translate(0, amount));
            }
        });
    }

    /*private shiftLastSelection(amount: number): void {
        const lastIndex = this.replaceRanges.length - 1;

        this.replaceRanges[lastIndex] = new Selection(this.replaceRanges[lastIndex].start.translate(0, amount), this.replaceRanges[lastIndex].end.translate(0, amount));
    }*/

    /**
     * Replaces text at `range` with `text`.
     *
     * _Does more than `TextEditorEdit.replace()`._
     *
     * @private
     * @param range Range at which text will be replaced.
     * @param text Text to replace with.
     * @returns true
     */
    private replace(range: Range, text: string): true {
        const textOld = this.vDoc.getText(range);
        const diffLength = text.length - textOld.length;

        // console.warn(`${text}'s diffLength = ${diffLength}`);

        this.edit.replace(range, text);

        let rangeNew = new Range(range.start, range.start.translate(0, text.length));

        if (this.hiddenSels.isOn && this.hiddenSels.pos.line === rangeNew.start.line && rangeNew.start.isAfter(this.hiddenSels.pos)) {
            rangeNew = new Range(rangeNew.start.translate(0, this.hiddenSels.diff), rangeNew.end.translate(0, this.hiddenSels.diff));
        }

        if (diffLength !== 0) {
            if (this.hasSelectionsAfter(range.end)) {
                // FIXME: All combinations work besides [1][0][2]
                // ^^^^ I THINK THIS IS FIXED ^^^^^
                this.shiftSelectionsAfter(range.end, diffLength);
            }
            if (this.hasHiddenSelsAfter(range.end)) {
                if (this.hiddenSels.isOn) {
                    this.hiddenSels.diff += diffLength;
                } else {
                    this.hiddenSels.diff = diffLength;
                }

                this.hiddenSels.isOn = true;
                this.hiddenSels.pos = range.end;
            }
        }

        if (this.isReversed) {
            this.replaceRanges.push(new Selection(rangeNew.end, rangeNew.start));
        } else {
            this.replaceRanges.push(new Selection(rangeNew.start, rangeNew.end));
        }

        this.lastRange = rangeNew;
        this.lastValueDiff = diffLength;

        return true;
    }

    /*private getPrevText(length: number, pos = this.wordRange.start) {
        let len = new BigNumber(length).abs();
        len = (len.isZero()) ? new BigNumber(1) : len;
        const range = new Range(pos.translate(0, len.negated().toNumber()), pos);

        return {
            range,
            text: this.vDoc.getText(range)
        };
    }*/

    /**
     * Check a line for a specific pattern and return it if it exists.
     *
     * @private
     * @param pos Line Position to check at.
     * @param pattern Pattern to search for.
     * @returns No matches: false; One match: {range, text}; Many matches: array of match objects;
     */
    /*private checkLineForPattern(pos: Position, pattern: RegExp) {
        const line = this.vDoc.lineAt(pos);
        const lineText = this.vDoc.getText(line.range);
        const results = [];

        if (pattern.test(lineText)) {
            const matches = lineText.match(pattern);

            for (const match of matches) {
                const startPos = line.range.start.translate(0, lineText.indexOf(match));
                const endPos = startPos.translate(0, match.length);

                results.push({
                    range: new Range(startPos, endPos),
                    text: match
                });
            }

            if (results.length > 1) {
                return results;
            } else {
                return results[0];
            }
        } else {
            return false;
        }
    }*/
}

// this method is called when your extension is deactivated
// export function deactivate() {
// }
